File: src/org/cappuccino_project/ide/intellij/plugin/indices/keys/StringPair.java
package org.cappuccino_project.ide.intellij.plugin.indices.keys;

import com.intellij.util.io.KeyDescriptor;
import org.jetbrains.annotations.NotNull;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;
import java.util.Objects;

public class StringPair {
    private String val1;
    private String val2;

    public StringPair(@NotNull final String className, @NotNull final String val2) {
        this.val1 = className;
        this.val2 = val2;
    }

    @NotNull
    public String getVal1() {
        return val1;
    }

    @NotNull
    public String getVal2() {
        return val2;
    }

    @Override
    public boolean equals(Object ob) {
        if (ob == this) {
            return true;
        }
        if (ob instanceof StringPair) {
            final StringPair classMethodPair = (StringPair)ob;
            return val1.equals(classMethodPair.val1) && val2.equals(classMethodPair.val2);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return Objects.hash(val1, val2);
    }

    public static final KeyDescriptor<StringPair> KEY_DESCRIPTOR = new KeyDescriptor<StringPair>() {
        @Override
        public int getHashCode(StringPair classMethodPair) {
            return classMethodPair.hashCode();
        }

        @Override
        public boolean isEqual(StringPair classMethodPair, StringPair t1) {
            return classMethodPair.equals(t1);
        }

        @Override
        public void save(
                @NotNull
                        DataOutput dataOutput, StringPair classMethodPair) throws IOException {
            dataOutput.writeUTF(classMethodPair.getVal1());
            dataOutput.writeUTF(classMethodPair.getVal2());
        }

        @Override
        public StringPair read(
                @NotNull
                        DataInput dataInput) throws IOException {
            final String val1 = dataInput.readUTF();
            final String val2 = dataInput.readUTF();
            return new StringPair(val1, val2);
        }
    };
}

File: src/org/cappuccino_project/ide/intellij/plugin/indices/keys/StringPair.kt
package org.cappuccino_project.ide.intellij.plugin.indices.keys

import com.intellij.util.io.KeyDescriptor

import java.io.DataInput
import java.io.DataOutput
import java.io.IOException
import java.util.Objects

class StringPair(val val1: String, val val2: String) {

    override fun equals(ob: Any?): Boolean {
        if (ob === this) {
            return true
        }
        if (ob is StringPair) {
            val classMethodPair = ob as StringPair?
            return val1 == classMethodPair!!.val1 && val2 == classMethodPair.val2
        }
        return false
    }

    override fun hashCode(): Int {
        return Objects.hash(val1, val2)
    }

    companion object {

        val KEY_DESCRIPTOR: KeyDescriptor<StringPair> = object : KeyDescriptor<StringPair> {
            override fun getHashCode(classMethodPair: StringPair): Int {
                return classMethodPair.hashCode()
            }

            override fun isEqual(classMethodPair: StringPair, t1: StringPair): Boolean {
                return classMethodPair == t1
            }

            @Throws(IOException::class)
            override fun save(
                    dataOutput: DataOutput, classMethodPair: StringPair) {
                dataOutput.writeUTF(classMethodPair.val1)
                dataOutput.writeUTF(classMethodPair.val2)
            }

            @Throws(IOException::class)
            override fun read(
                    dataInput: DataInput): StringPair {
                val val1 = dataInput.readUTF()
                val val2 = dataInput.readUTF()
                return StringPair(val1, val2)
            }
        }
    }
}
